### PR TITLE
Week 5: Virtuoso Hybrid Store

### PR DESCRIPTION
**Pull request by shubh-ranade**

### CODE REVIEW

File: index.md [hunk 1]
```diff
@@ -16,7 +16,8 @@ layout: default
     - [DB architecture Section 6, 7](./week4/db-architecture-6-7.md)
     - [Note on Jive join](./week4/jive-join.md)
 * Week 5
-    - [Paper 3: Columnar Storage in Microsoft SQL Server 2012](./week5/3_columnar_storage_in_sql_server_12.md)
```
Comment (neelkshah) [inline index.md]: Awesome! This was low key irritating

**Review by neelkshah — APPROVED** [review]

File: week5/1-virtuoso-hybrid-store.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
+---
+layout: default
+---
+
+# Virtuoso, a Hybrid RDBMS/Graph Column Store
+
+___
+
+This paper discusses the design choices met in applying column store techniques under the twin requirements of performing well on the unpredictable, semi-structured RDF data and more typical relational BI workloads. Virtuoso started out as a row-wise transaction oriented RDBMS and transitioned to being an RDF graph store and subsequently started using column-wise storage and vectored execution. The main incentive for using column-wise storage was its excellent space efficiency, which suited Virtuoso's large RDF applications.
```
Comment (twitu) [inline week5/1-virtuoso-hybrid-store.md]: `RDF` I don't know what this means. Consider adding a full form first before using its acronym.

File: week5/1-virtuoso-hybrid-store.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+
+## Column Store Implementation
+
+* Virtuoso supports both row-wise and column-wise stored tables. The table is simply the index (B-tree) on its primary key with the dependent part following the key on the index leaf. The table is sorted on its primary key.
+
+* For row-wise indexes, the table row itself is stored on the index leaf corresponding to the primary key entry.
+
+* In case of column-wise indexes, the table is divided into segments (horizontal partitions, thousands of rows each) and the index is sparse and row-wise at the top, with the leaf storing references to compressed columns of that segment.
```
Comment (twitu) [inline week5/1-virtuoso-hybrid-store.md:18]: An image if possible.

File: week5/1-virtuoso-hybrid-store.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+* Virtuoso has a small page size (8K) as compared to other column stores to facilitate the storage of both row-wise and column-wise data structures in the same buffer pool. Smaller pages also means low latency for random inserts.
```
Comment (twitu) [inline week5/1-virtuoso-hybrid-store.md:22]: Really nice.

File: week5/1-virtuoso-hybrid-store.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+* Compression schemes are column dependent and auto-tuned by the database.
+
+* Virtuoso doesn't have the concept of a table wide row number (row id). Instead, rows have value based keys which can be used as partitioning keys for large cluster-based databases. Different indexes can be partitioned according to different columns and even stored on different nodes.
+
+
+## Vectored Execution
+
+* Benefits of columnar vectored execution (improved cache locality, tight loops, improved utilization of CPU memory throughput) were found to be applicable to row store performance as well.
+
+* Pipelined operators like join, which change cardinality (no. of rows) of the output and add columns to the output, maintain input to output row mappings so that tuples can be reconstructed at the end. Reconstructing a large number of rows together optimizes memory bandwidth.
```
Comment (twitu) [inline week5/1-virtuoso-hybrid-store.md:33]: Is this similar to jive join or later materialization? Seems like similar techniques are popping up.

File: week5/1-virtuoso-hybrid-store.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+* Virtuoso uses vectoring for index lookups too. Since density of hits in an index determines performance, long vectors are preferred, because new lookups don't have to be made frequently as long as the value to be fetched is in the same segment (or same row-wise leaf page). This is especially useful for large read queries.
```
Comment (twitu) [inline week5/1-virtuoso-hybrid-store.md:35]: But isn't the page size small (8k), so how can vectors be long? Or am I getting it wrong?

**Review by twitu — COMMENTED** [review]
Will you be explaining these in the next meetup?

File: week5/1-virtuoso-hybrid-store.md (hunk 6 above)
Comment (shubh-ranade) [reply week5/1-virtuoso-hybrid-store.md:35]: Vectors don't need to fit exactly in one page.

Comment (twitu) [reply week5/1-virtuoso-hybrid-store.md:35]: But then what does long vector mean. I always assumed a vector to be an in-memory construct.

File: week5/1-virtuoso-hybrid-store.md (hunk 5 above)
Comment (shubh-ranade) [reply week5/1-virtuoso-hybrid-store.md:33]: The mappings don't have to be related to the join algorithm used. The paper doesn't discuss any particular algorithm. For example, jive join uses a join index and there's no mention of join indexes.

**Comment by shubh-ranade** [comment]
Added an image

**Review by twitu — APPROVED** [review]

**Pull request merged**